### PR TITLE
Fix weight decay filtering for accuracy-compatible optimizer params

### PR DESCRIPTION
**Pull request by zhengshengning**
#### Before submitting

- [ ] Lint code. If there are lint issues, please format the code first.

```shell
# Install and register `pre-commit` in the project folder
pip install pre-commit && pre-commit install

# Process previous code files separately
pre-commit run --file XXXX.py
```

- [ ] Add test cases into `tests` folder. If there are codecov issues, please add tests cases first.

### PR types

New features

### PR changes

Others

### Description

Fix weight decay filtering for accuracy-compatible optimizer params
问题10: 优化器 对所有 len(shape)==1 的参数（norm/bias）设 wd_mult=0

### CODE REVIEW

File: paddleformers/trainer/trainer.py [hunk 1]
```diff
@@ -3412,17 +3412,40 @@ def apply_decay_param_fun(x):
...
 
-            if self.optimizer_grouped_parameters is not None:
-                params = self.optimizer_grouped_parameters
-                # A plain list only customizes the trainable set, so it should still use the default decay filter.
-                # But dict may define per-group weight_decay explicitly, so do not override.
-                is_param_group_dict = (
-                    isinstance(params, (list, tuple)) and len(params) > 0 and isinstance(params[0], dict)
-                )
-                apply_decay_param_fun = None if is_param_group_dict else _build_apply_decay_param_fun()
+            if getattr(self.model.config, "use_accuracy_compatible", False):
```
Comment (risemeup1111) [inline paddleformers/trainer/trainer.py]: P1 **优先级：P1**

**问题：** 这里直接访问 `self.model.config`，会让 `Trainer` 仍支持的普通 `paddle.nn.Layer` 在创建 optimizer 时抛 `AttributeError`。`Trainer` 文档允许 `model` 是 `PretrainedModel` 或 `paddle.nn.Layer`，并且 `__init__` 里也是用 `getattr(self.model, "config", None)` 做可选处理；本次新增的无保护访问会在默认 AdamW 路径提前触发。

**影响：** 没有 `config` 的自定义模型即使不使用 accuracy-compatible 模式，也会在 `create_optimizer()` 里失败。

**处理要求：请针对该评论修复并提交新的 commit。**

**建议：** 先把 `config` 当作可选属性读取：
```suggestion
            if getattr(getattr(self.model, "config", None), "use_accuracy_compatible", False):
```

File: paddleformers/trainer/trainer.py [hunk 2]
```diff
@@ -3412,17 +3412,40 @@ def apply_decay_param_fun(x):
...
+                if isinstance(params, list) and params and isinstance(params[0], dict):
+                    apply_decay_param_fun = None
+                else:
+                    optimizer_param_ids = {id(p) for p in params}
+                    decay_parameters = {
+                        p.name
+                        for n, p in self.model.named_parameters()
+                        if id(p) in optimizer_param_ids
+                        and not p.stop_gradient
+                        and not any(nd in n for nd in ["bias", "norm"])
```
Comment (risemeup1111) [inline paddleformers/trainer/trainer.py]: P1 **优先级：P1**

**问题：** 新增注释说明 accuracy-compatible 模式要对齐 Megatron 的 `len(shape)==1` 不做 weight decay，但这里仍按参数名包含 `bias/norm` 过滤。这样一维但名称不含这两个片段的参数仍会进入 `decay_parameters`，例如仓库里的 `Phi4DiffAttention.lambda_q1/lambda_k1/...` 和 `Phi4MambaMixer.D` 都是一维可训练参数。

**影响：** 开启 `use_accuracy_compatible` 后，这类一维参数仍会被 AdamW 衰减，实际行为和新增注释里的 Megatron 对齐规则不一致，仍可能造成精度/权重 md5 漂移。

**处理要求：请针对该评论修复并提交新的 commit。**

**建议：** 在 accuracy-compatible 分支按参数 shape 过滤：
```suggestion
                        and len(p.shape) > 1
```

File: paddleformers/trainer/trainer.py [hunk 3]
```diff
@@ -3412,17 +3412,40 @@ def apply_decay_param_fun(x):
...
+                else:
+                    params = [p for p in self.model.parameters() if not p.stop_gradient]
+
+                # Align with Megatron: MG sets wd_mult=0 for all params with len(shape)==1 (norm/bias),
+                # i.e. no weight decay. When PF SFT/DPO passes a flat Parameter list via
+                # set_optimizer_grouped_parameters, the original logic set apply_decay_param_fun to None,
+                # causing norm/bias to also be weight-decayed. The fp32 master then drifts step by step
+                # from step1 (bf16 masks it until the weight md5 diverges at step4). Only keep None when
+                # params is already a dict list with weight_decay groups (the groups carry their own wd config).
+                if isinstance(params, list) and params and isinstance(params[0], dict):
```
Comment (risemeup1111) [inline paddleformers/trainer/trainer.py]: P2 **优先级：P2**

**问题：** 这里把 param-group dict 的判断缩窄成了 `list`，但旧实现显式支持 `(list, tuple)`，`set_optimizer_grouped_parameters()` 的文档也没有限制只能传 list。若调用方传入 tuple 形式的 param groups，会落入下面的 flat-parameter 分支，`{id(p) for p in params}` 收集到的是 dict 的 id，最终 `decay_parameters` 为空。

**影响：** 这种兼容用法下 `apply_decay_param_fun` 会对真实参数全部返回 `False`，从而覆盖参数组里的 weight decay 语义。

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

**建议：** 保持旧实现已支持的 tuple 形态；如果确认新接口不再支持 tuple，请回复说明兼容性取舍：
```suggestion
                if isinstance(params, (list, tuple)) and len(params) > 0 and isinstance(params[0], dict):
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
发现几处需要先修复的兼容性和准确性问题，细节已放在行级评论里。当前还有多项 CI 仍在运行中，合入前也需要等待完整结果。

P3 **优先级：P3**

非行级：PR 标题和描述不属于 changed diff 行，无法行级附着。

**问题：** 当前标题和描述都只有 “Align with Megatron”，没有说明实际变更范围、触发场景和验证方式。

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

**建议：** 标题可改成类似 `Fix weight decay filtering for accuracy-compatible optimizer params`，描述中补充 `use_accuracy_compatible`、`optimizer_grouped_parameters` 的触发路径，以及已运行或等待中的精度/训练验证结果。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/trainer/trainer.py [hunk 4]
```diff
@@ -3412,17 +3412,38 @@ def apply_decay_param_fun(x):
...
+
             else:
-                params = [p for p in self.model.parameters() if not p.stop_gradient]
-                apply_decay_param_fun = _build_apply_decay_param_fun()
+                if self.optimizer_grouped_parameters is not None:
+                    params = self.optimizer_grouped_parameters
+                    apply_decay_param_fun = None
+                else:
+                    params = [p for p in self.model.parameters() if not p.stop_gradient]
+                    apply_decay_param_fun = _build_apply_decay_param_fun()
```
Comment (risemeup1111) [inline paddleformers/trainer/trainer.py:3449]: P1 **优先级：P1**

**问题：** `use_accuracy_compatible` 关闭时，这个分支现在对任意 `optimizer_grouped_parameters` 都把 `apply_decay_param_fun` 设为 `None`。但左侧旧逻辑只在 param-group dict 时禁用 decay filter，flat 参数列表仍会使用 `_build_apply_decay_param_fun()` 排除 bias/norm。SFT/DPO workflow 当前会通过 `set_optimizer_grouped_parameters(trainable_parameters)` 传入 flat 参数列表，因此默认配置下会把 bias/norm 也纳入 weight decay。

**影响：** 这是默认路径的行为回归；未开启 accuracy-compatible 的常规训练也会改变权重衰减语义，可能影响训练精度和既有结果可复现性。

**处理要求：请针对该评论修复并提交新的 commit。**

**建议：** 非 accuracy-compatible 分支保持旧的 param-group dict 判断，只在 dict groups 自带 weight decay 配置时设为 `None`：
```suggestion
                if self.optimizer_grouped_parameters is not None:
                    params = self.optimizer_grouped_parameters
                    is_param_group_dict = (
                        isinstance(params, (list, tuple)) and len(params) > 0 and isinstance(params[0], dict)
                    )
                    apply_decay_param_fun = None if is_param_group_dict else _build_apply_decay_param_fun()
                else:
                    params = [p for p in self.model.parameters() if not p.stop_gradient]
                    apply_decay_param_fun = _build_apply_decay_param_fun()
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交。此前三条行级代码问题在当前 head 上已修复，但又发现一处默认训练路径的 weight decay 行为回归，细节已放在新的行级评论里；合入前还需要等待当前 CI 完整结束。

P3 **优先级：P3**

非行级：PR 描述不属于 changed diff 行，无法行级附着。

**问题：** 标题已经更新，但 PR 描述仍只有 “Align with Megatron”，还缺少 `use_accuracy_compatible`、`optimizer_grouped_parameters` 触发路径和验证结果说明。

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

**建议：** 在描述里补充本 PR 修改的 optimizer decay 行为、SFT/DPO flat 参数列表场景，以及已运行或待完成的精度/训练验证。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — COMMENTED** [review]
已复查最新提交，上轮行级 P1 问题在当前 head 上已经修复，暂未发现新的代码阻塞项。当前 CI 仍在运行，请合入前等待完整结果。

P3 **优先级：P3**

非行级：PR 描述不属于 changed diff 行，无法行级附着。

**问题：** PR 标题已经更新，但描述仍只有 “Align with Megatron”，还缺少 `use_accuracy_compatible`、`optimizer_grouped_parameters` 触发路径和验证结果说明。

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

**建议：** 请在描述里补充本 PR 修改的 optimizer weight decay 行为、SFT/DPO flat 参数列表场景，以及已运行或待完成的精度/训练验证。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by DanielSun11 — APPROVED** [review]
